### PR TITLE
[noticket] s3

### PR DESCRIPTION
**Pull request by sk-jlewis**

### CODE REVIEW

**Review by paddycarey — COMMENTED** [review]
This seems good to me, but my ruby-fu is weak, so I'd like some rubyists to review if possible. @TheTeaNerd or @deanobarnett perhaps?

File: lib/gemstash/s3.rb [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+    # If the storage engine detects the base cache directory was originally
+    # initialized with a newer version, this error is thrown.
+    class VersionTooNew < StandardError
+      def initialize(folder, version)
+        super("Gemstash storage version #{Gemstash::S3::VERSION} does " \
+              "not support version #{version} found at #{folder}")
+      end
+    end
+
+    # This object should not be constructed directly, but instead via
```
Comment (TheTeaNerd) [inline lib/gemstash/s3.rb:25]: So make it private

**Review by TheTeaNerd — APPROVED** [review]
There are issues like https://github.com/bundler/gemstash/issues/107 discussing S3 storage; should this PR be pushed to the parent project?

The Ruby is fine, looks good. I'm a bit unsure about the cost/benefit of writing our own _vs_ using a SaaS provider, but if this works, I'm :+1: to :shipit:

**Comment by sk-jlewis** [comment]
> There are issues like bundler #107 discussing S3 storage; should this PR be pushed to the parent project?

yup!

File: gemstash.gemspec [hunk 2]
```diff
@@ -38,6 +38,8 @@ you push your own private gems as well."
   spec.add_runtime_dependency "sequel", "~> 4.26"
   spec.add_runtime_dependency "sinatra", "~> 1.4"
   spec.add_runtime_dependency "thor", "~> 0.19"
+  spec.add_runtime_dependency "aws-sdk", "~> 2"
+  spec.add_runtime_dependency "byebug"
```
Comment (deanobarnett) [inline gemstash.gemspec:42]: Can we remove this since it's a dev dependancy?

File: lib/gemstash/s3.rb [hunk 3]
```diff
@@ -0,0 +1,345 @@
+require "gemstash"
+require "digest"
+require "fileutils"
+require "pathname"
+require "yaml"
+require "aws-sdk"
+
+module Gemstash
```
Comment (deanobarnett) [inline lib/gemstash/s3.rb:8]: There's a lot going on in this class. Could we move out some of the different parts into separate files?

File: man/gemstash-configuration.5.md [hunk 4]
```diff
@@ -47,6 +49,26 @@ ahead of time and be writable to the Gemstash server process. Specifying the
 
 Any valid path
 
+# Storage Adapter
+
+`:storage_adapter`
+
+Specifies how gems should be stores
```
Comment (deanobarnett) [inline man/gemstash-configuration.5.md:56]: `stores` => `stored`

**Review by kaycoinsUSA — COMMENTED** [review]
👍

**Review by kaycoinsUSA — COMMENTED** [review]
Looks good to me.

**Pull request merged**